pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,913 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multishare_reconciler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	filev1beta1 "google.golang.org/api/file/v1beta1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	v1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apimachinery/pkg/util/wait"
+	storageListers "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/gcp-filestore-csi-driver/pkg/apis/multishare/v1alpha1"
+	clientset "sigs.k8s.io/gcp-filestore-csi-driver/pkg/client/clientset/versioned"
+	informers "sigs.k8s.io/gcp-filestore-csi-driver/pkg/client/informers/externalversions/multishare/v1alpha1"
+	listers "sigs.k8s.io/gcp-filestore-csi-driver/pkg/client/listers/multishare/v1alpha1"
+	cloud "sigs.k8s.io/gcp-filestore-csi-driver/pkg/cloud_provider"
+	"sigs.k8s.io/gcp-filestore-csi-driver/pkg/cloud_provider/file"
+	driver "sigs.k8s.io/gcp-filestore-csi-driver/pkg/csi_driver"
+	"sigs.k8s.io/gcp-filestore-csi-driver/pkg/util"
+)
+
+type Op struct {
+	Id     string
+	Type   util.OperationType
+	Target string
+	Err    error
+}
+
+type multishareReconciler struct {
+	clientset clientset.Interface
+	config    *driver.GCFSDriverConfig
+	cloud     *cloud.Cloud
+
+	shareLister       listers.ShareInfoLister
+	shareListerSynced cache.InformerSynced
+
+	instanceLister       listers.InstanceInfoLister
+	instanceListerSynced cache.InformerSynced
+
+	scLister storageListers.StorageClassLister
+}
+
+func NewMultishareReconciler(
+	clientset clientset.Interface,
+	config *driver.GCFSDriverConfig,
+	shareInformer informers.ShareInfoInformer,
+	instanceInformar informers.InstanceInfoInformer,
+	scLister storageListers.StorageClassLister,
+) *multishareReconciler {
+	recon := &multishareReconciler{
+		clientset: clientset,
+		cloud:     config.Cloud,
+		config:    config,
+		scLister:  scLister,
+	}
+
+	recon.shareLister = shareInformer.Lister()
+	recon.shareListerSynced = shareInformer.Informer().HasSynced
+
+	recon.instanceLister = instanceInformar.Lister()
+	recon.instanceListerSynced = instanceInformar.Informer().HasSynced
+
+	return recon
+}
+
+func (recon *multishareReconciler) Run(stopCh <-chan struct{}) {
+	klog.Infof("Starting multishare reconciler")
+	defer klog.Infof("Shutting down multishare reconciler")
+
+	informerSynced := []cache.InformerSynced{recon.shareListerSynced, recon.instanceListerSynced}
+	if !cache.WaitForCacheSync(stopCh, informerSynced...) {
+		klog.Errorf("Cannot sync caches")
```

Comment:
Should we fatal here? 
If cache sync fails, and it fails silently and returns (thereby does not start the reconciler worker). Now the CSI calls would stall, because there is no reconciler to react to the CRDs.

Reply:
I wonder if the driver should crash here? although it might send the whole component into crashLooping.

logging worker start is definitely a great point.

Reply:
yeah let's not crash the entire driver. In L95 we should log something like "starting cache sync ..." also.

Reply:
done.